### DISCUSSION TITLE
How can I get a usable authorization_code bearer token with Push Authorization Request disabled?

### DISCUSSION THREAD

**Discussion by LePau** [authelia/authelia#7536] [Troubleshooting] [answered]
I'm trying to set up an authorization_code flow using [HttpYac](https://httpyac.github.io/).  I set up the authelia config based on the [Authorization Code Flow](https://www.authelia.com/integration/openid-connect/oauth-2.0-bearer-token-usage/#examples) snippet from the docs.

However, I receive an error response from Authelia regarding Push Authorization Requests (PARs):

```
{
    "error": "invalid_request",
    "error_description": "The request is missing a required parameter, includes an invalid parameter value, includes a parameter more than once, or is otherwise malformed. Pushed Authorization Requests are required but this Authorization Request was not made as a Pushed Authorization Request. The registered OAuth 2.0 client with id 'localhost3000_2024_07_14_authorization_code' is registered with a policy which requires Pushed Authorization Requests be used."
}
```

According to the [docs](https://www.authelia.com/configuration/identity-providers/openid-connect/clients/#require_pushed_authorization_requests):

> Important Note: A majority of clients will not support this option as it requires a special but highly secure authorization flow.

Looks like HttpYac doesn't support PAR.  OK, so I disabled it:

```
    - client_id: localhost3000_2024_07_14_authorization_code
    ...    
    scopes:
        - offline_access
        - authelia.bearer.authz
    ...
    require_pushed_authorization_requests: false
```

After redeploy and rerun of auth code flow, Authelia fails to start and logs say:

> time="..." level=error msg="Configuration: identity_providers: oidc: clients: client 'localhost3000_2024_07_14_authorization_code': option 'require_pushed_authorization_requests' must be configured as 'true' when configured with scope 'authelia.bearer.authz' but it's configured as 'false'"

I remove the `authelia.bearer.authz` scope:

```
    - client_id: localhost3000_2024_07_14_authorization_code
    ...    
    scopes:
        - offline_access
    ...
    require_pushed_authorization_requests: false
```

Then, I re-deployed authelia and re-ran the authorization code flow.  I get a response with an access_token this time.  Note I'm using the `offline_access` scope, but I've used the `openid` scope with the same results:

Request:
```
POST https://localhost.localdomain:8080/auth/api/oidc/token
accept-encoding: gzip, deflate, br
authorization: Basic b...c=
content-length: 288
content-type: application/x-www-form-urlencoded
user-agent: got (https://github.com/sindresorhus/got)

grant_type=authorization_code&scope=offline_access&code=authelia_ac_...x2M&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fcallback&code_verifier=AX...i58  
```

Response:

```
{
    "access_token": "authelia_at_yT6...-5o",
    "expires_in": 3600,
    "refresh_token": "authelia_rt_3s..._yMg",
    "scope": "offline_access",
    "token_type": "bearer"
}
```

However when I try to use the access token I get an error.

Request:

```
GET https://localhost.localdomain:8080/protected
accept: */*  
accept-encoding: gzip, deflate, br  
content-type: application/json  
proxy-authorizatio*: bearer authelia_at_yT6Z...X-5o  
user-agent: httpyac  
```

Response:
```
HTTP/1.1 302 - Moved Temporarily
connection: close  
content-length: 145  
content-type: text/html  
date: Wed, 17 Jul 2024 17:09:18 GMT  
location:   
server: nginx/1.21.1  
www-authenticate: Bearer error="invalid_token",error_description="The access token is expired, revoked, malformed, or invalid for other reasons. The client can obtain a new access token and try again."  
```

The logs show that the `authelia.bearer.authz` scope is required in the token (and implicitely the client):

> time="..." level=debug msg="Error occurred while attempting to authenticate a request but the matched rule was a bypass rule" caller="github.com/authelia/authelia/v4/internal/handlers/handler_authz.go:91 (*Authz).Handler" error="failed to validate Proxy-Authorization header with bearer scheme: error performing token introspection: The requested scope is invalid, unknown, or malformed. The request scope 'authelia.bearer.authz' has not been granted or is not allowed to be requested." method=GET path=/api/authz/auth-request remote_ip=127.0.0.1

To recap:

* I can retrieve an access_token using authorization_code grant_type client.
* However, I can't use the token because Authelia requires it to have the `authelia.bearer.authz` scope.
* I can't add the scope to the request, because the client doesn't have that scope defined.
* I can't define that scope for the client because Authelia requires `require_pushed_authorization_requests` set to true.
* I can't set `require_pushed_authorization_requests` to true, because the http client (HttpYac) doesn't support PAR.

Is there a way to disable PAR and request a usable bearer access_token using the authorization_code grant type?

<details>
  <summary>Full config</summary>
<pre>
theme: light
default_2fa_method: ''
server:
    # host: 0.0.0.0
    # port: 9091
    address: tcp://:9091/auth
    asset_path: ''
    headers:
    csp_template: ''
    buffers:
    read: 4096
    write: 4096
    timeouts:
    read: 6 seconds
    write: 6 seconds
    idle: 30 seconds
    endpoints:
    authz:
        auth-request:
        implementation: 'AuthRequest'
        authn_strategies:
            - name: 'HeaderAuthRequestProxyAuthorization'
            schemes:
                - 'Bearer'
            - name: 'CookieSession'
    enable_pprof: false
    enable_expvars: false
log:
    level: trace
    format: text
    file_path: ''
    keep_stdout: true
identity_providers:
    oidc:
    enable_client_debug_messages: true
    # hmac_secret: 'IVPWBkAdJHje3uz7LtFTDU2pFUfh39Xm' # covered by env var
    jwks:
        - key: {{ secret "/secrets/AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_KEY" | mindent 14 "|" | msquote }}
        certificate_chain: {{ secret "/secrets/AUTHELIA_IDENTITY_PROVIDERS_OIDC_JWKS_CERTIFICATE_CHAIN" | mindent 14 "|" | msquote }}
    clients:
        - client_id: $CLUSTER_AUTH_LOCAL_CLIENT_CREDENTIALS_ID
        client_secret: {{ secret "/secrets/CLUSTER_AUTH_LOCAL_CLIENT_CREDENTIALS_SECRET_HASHED" | msquote }} 
        client_name: clientcredentials
        public: false
        require_pkce: true
        pkce_challenge_method: 'S256'
        consent_mode: 'explicit'
        # authorization_policy: 'one_factor' # client_credentials already one factor by default
        audience:
            - ${CLUSTER_URL}
        scopes:
            - authelia.bearer.authz
        token_endpoint_auth_method: 'client_secret_basic'
        grant_types:
            - client_credentials
        - client_id: localhost3000_2024_07_14_authorization_code
        client_secret: {{ secret "/secrets/CLUSTER_AUTH_AUTHORIZATION_CODE_CLIENT_SECRET_HASHED" | msquote }} 
        public: false
        require_pkce: true
        pkce_challenge_method: 'S256'
        redirect_uris:
            # yttpyac for 9999
            - http://localhost:9999/callback
            - $CLUSTER_URL/argo
        scopes:
            - offline_access
        authorization_policy: one_factor              
        audience:
            - $CLUSTER_URL
        grant_types:
            - 'authorization_code'
            - 'refresh_token'
        response_types:
            - 'code'
        response_modes:
            - 'form_post'
        consent_mode: 'explicit'
        require_pushed_authorization_requests: false
        token_endpoint_auth_method: 'client_secret_basic'              
totp:
    disable: false
    issuer: Authelia
    algorithm: SHA1
    digits: 6
    period: 30
    skew: 1
    secret_size: 32
webauthn:
    disable: false
    display_name: Authelia
    attestation_conveyance_preference: indirect
    user_verification:
    preferred
    timeout: 60 seconds
ntp:
    address: udp://time.cloudflare.com:123
    version: 4
    max_desync: 3 seconds
    disable_startup_check: false
    disable_failure: false
authentication_backend:
    password_reset:
    disable: false
    custom_url: ''
    file:
    path: /config/users_database.yaml
    watch: false
    search:
        email: false
        case_insensitive: false
    password:
        algorithm: 'argon2'
        argon2:
        variant: 'argon2id'
        iterations: 3
        memory: 65536
        parallelism: 4
        key_length: 32
        salt_length: 16
        scrypt:
        iterations: 16
        block_size: 8
        parallelism: 1
        key_length: 32
        salt_length: 16
        pbkdf2:
        variant: 'sha512'
        iterations: 310000
        salt_length: 16
        sha2crypt:
        variant: 'sha512'
        iterations: 50000
        salt_length: 16
        bcrypt:
        variant: 'standard'
        cost: 12
password_policy:
    standard:
    enabled: false
    min_length: 8
    max_length: 0
    require_uppercase: true
    require_lowercase: true
    require_number: true
    require_special: true
    zxcvbn:
    enabled: false
    min_score: 0
session:
    remember_me: 1 month
    cookies:
    - name: 'authelia_session'
        domain: ${CLUSTER_HOSTNAME}
        authelia_url: ${CLUSTER_AUTH_URL}
        default_redirection_url: ${CLUSTER_URL}
        same_site: 'lax'
        expiration: 1 hour
        inactivity: 5 minutes
regulation: 
    ban_time: 5 minutes
    find_time: 2 minutes
    max_retries: 3
storage:
    local:
    path: /config/db.sqlite3
notifier:
    disable_startup_check: false
    filesystem:
    filename: /config/notification.txt
access_control:
    default_policy: deny
    rules:
    - domain: '${CLUSTER_HOSTNAME}'
        policy: one_factor
        subject:
        - user:authelia
        - oauth2:client:$CLUSTER_AUTH_LOCAL_CLIENT_CREDENTIALS_ID
</pre>
</details>

**Comment by james-d-elliott** [comment, marked as answer]
You can either use the pushed authorization request flow or client credentials flow to get a bearer token with that scope, we don't support it any other way.

**Comment by james-d-elliott** [comment]
Forgot to link this: https://www.authelia.com/integration/openid-connect/oauth-2.0-bearer-token-usage/
